src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,725 @@
+    const struct iovec iovec = {
+        .iov_base = (void*)ping_str,
+        .iov_len  = sizeof(ping_str),
+    };
+    return perftest_mad_sendv(rte_group, &iovec, 1);
+}
+
+static int perftest_mad_is_ping(void *buffer, size_t size)
+{
+    size_t ping_size = sizeof(ping_str);
+    if (size != ping_size) {
+        return 0;
+    }
+    return memcmp(buffer, ping_str, ping_size) == 0;
+}
+
+static ucs_status_t perftest_mad_recv(perftest_mad_rte_group_t *rte_group,
+                                      void *buffer, size_t *avail,
+                                      ib_portid_t *remote_port)
+{
+    int fd         = mad_rpc_portid(rte_group->mad_port);
+    size_t ref_len = *avail + IB_VENDOR_RANGE2_DATA_OFFS;
+    void *umad     = NULL;
+    int ret;
+    uint8_t *data;
+    int len; /* cannot use 'size_t' here */
+    struct ib_user_mad *user_mad;
+    ucs_status_t status;
+
+    for (;;) {
+        if ((ref_len > (INT_MAX - umad_size())) || (ref_len == 0)) {
+            ucs_error("MAD: receive: invalid umad_recv length %zu", ref_len);
+            status = UCS_ERR_INVALID_PARAM;
+            goto terminate;
+        }
+
+        umad = realloc_or_free(umad, umad_size() + ref_len);
+        if (umad == NULL) {
+            return UCS_ERR_NO_MEMORY;
+        }
+
+    ping_retry:
+        user_mad = umad;
+        len      = ref_len;
+
+        ret = umad_recv(fd, umad, &len, 3 * 1000);
+        if (ret >= 0) {
+            break;
+        }
+
+        if ((errno == ETIMEDOUT) || (errno == EAGAIN)) {
+            status = UCS_ERR_NO_PROGRESS;
+            goto terminate;
+        } else if ((errno != EINVAL) && (errno != ENOSPC)) {
+            ucs_error("MAD: umad_recv(fd %d, len %d) ret:%d failed: %m", fd,
+                      len, ret);
+            status = UCS_ERR_IO_ERROR;
+            goto terminate;
+        } else if (errno == EINVAL) {
+            /* EINVAL: no length info */
+            ref_len *= 2; /* 'len' has become invalid */
+        } else {
+            /* ENOSPC: needed length is set */
+            ucs_assert(errno == ENOSPC);
+            ref_len = len;
+        }
+    }
+
+    if (perftest_mad_get_remote_port(umad, remote_port) != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto terminate;
+    }
+
+    len -= IB_VENDOR_RANGE2_DATA_OFFS;
+    if (len <= 0) {
+        if (user_mad->status == ETIMEDOUT) {
+            ucs_error("MAD: receive: Remote unreachable");
+            status = UCS_ERR_UNREACHABLE;
+        } else {
+            status = UCS_ERR_OUT_OF_RANGE;
+        }
+        goto terminate;
+    }
+
+    ret = umad_status(umad);
+    if (ret) {
+        ucs_warn("MAD: receive: Status failure: %d", ret);
+        status = UCS_ERR_REJECTED;
+        goto terminate;
+    }
+
+    status = UCS_OK;
+
+    data = umad_get_mad(umad);
+    data = UCS_PTR_BYTE_OFFSET(data, IB_VENDOR_RANGE2_DATA_OFFS);
+    if (perftest_mad_is_ping(data, len)) {
+        ucs_debug("PONG");
+        goto ping_retry;
```

Comment:
need to restructure this whole function logic, maybe use helper functions.
should not jump into the middle of the previous loop

Reply:
makes sense, added helper functions containing for(), kept label, retested with ka/ping, error handling on socket, mpi and mad rte on ucp test with/without threads